File: crates/prolly/bindings/ruby/examples/diff_merge.rb
# frozen_string_literal: true

require 'prolly'

engine = Prolly::ProllyEngine.memory(Prolly.default_config)
base = engine.create
base = engine.put(base, 'doc:title'.b, 'Draft'.b)

left = engine.put(base, 'doc:body'.b, 'Hello'.b)
right = engine.put(base, 'doc:tags'.b, 'example'.b)

left_changes = engine.diff(base, left)
raise 'expected one left-side change' unless left_changes.length == 1
raise 'expected doc:body diff' unless left_changes.first.key == 'doc:body'.b

merged = engine.merge(base, left, right, 'prefer_right')
raise 'missing body' unless engine.get(merged, 'doc:body'.b) == 'Hello'.b
raise 'missing tags' unless engine.get(merged, 'doc:tags'.b) == 'example'.b

puts "diff_merge: merged #{left_changes.length} left-side change"

File: crates/prolly/bindings/python/examples/diff_merge.py
from __future__ import annotations

import prolly


engine = prolly.ProllyEngine.memory(prolly.default_config())

base = engine.create()
base = engine.put(base, b"doc:title", b"Draft")

left = engine.put(base, b"doc:body", b"Hello")
right = engine.put(base, b"doc:tags", b"example")

left_changes = engine.diff(base, left)
assert len(left_changes) == 1
assert left_changes[0].key == b"doc:body"

merged = engine.merge(base, left, right, "prefer_right")
assert engine.get(merged, b"doc:body") == b"Hello"
assert engine.get(merged, b"doc:tags") == b"example"

print(f"diff_merge: merged {len(left_changes)} left-side change")
